### PR TITLE
rpc: return error code for eth_getWork when no work ready

### PR DESCRIPTION
**Pull request by Gustav-Simonsson**

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Wed Aug 26 10:47:17 UTC 2015_

**Comment by maran** [comment]
:+1:

File: miner/remote_agent.go [hunk 1]
```diff
@@ -110,9 +111,10 @@ func (a *RemoteAgent) GetWork() [3]string {
 		res[2] = common.BytesToHash(n.Bytes()).Hex()
 
 		a.work[block.HashNoNonce()] = a.currentWork
+		return res, nil
+	} else {
```
Comment (bas-vk) [inline miner/remote_agent.go]: You can remove the else.

File: rpc/api/eth.go [hunk 2]
```diff
@@ -563,7 +563,12 @@ func (self *ethApi) GetLogs(req *shared.Request) (interface{}, error) {
 
 func (self *ethApi) GetWork(req *shared.Request) (interface{}, error) {
 	self.xeth.SetMining(true, 0)
-	return self.xeth.RemoteMining().GetWork(), nil
+	ret, err := self.xeth.RemoteMining().GetWork()
+	if err != nil {
+		return nil, shared.NewNotReadyError("getWork")
```
Comment (bas-vk) [inline rpc/api/eth.go]: I'm not really sure if this is the most convenient/informative.
The function `GetWork` is actually ready but there is no work to return.

Comment (Gustav-Simonsson) [inline rpc/api/eth.go]: What string what would you suggest to return instead?

**Comment by fjl** [comment]
Maybe it should wait for work to appear instead (timing out after a second or so).

**Comment by Gustav-Simonsson** [comment]
@fjl Problem with adding waiting with a timeout is that it could be a very long time if there is a full chain sync on a slower machine. In that case a caller would still need to retry (unless the timeout is like 20 minutes or so), so I would keep it simple and let the caller decide on retries.

**Comment by fjl** [comment]
Sure, it was just a suggestion.

File: rpc/api/eth.go (hunk 2 above)
Comment (fjl) [inline rpc/api/eth.go]: "mining work"

**Comment by fjl** [comment]
We can merge this if you take care of Bas' comments.

**Comment by bas-vk** [comment]
:+1:

**Comment by sammy007** [comment]
While testing develop branch with this PR applied I noticed that daemon responds with work while it still catching up. Is it expected behaviour? I think it's ok because actually node is always catching up, it's impossible to predict that node is 100% in sync, but it worth posting here.

    I0829 19:33:25.173213   15751 chain_manager.go:661] imported 256 block(s) (0 queued 0 ignored) including 62 txs in 3.00820678s. #158750 [e79c44dd / f0ca3c3c]
    I0829 19:33:26.554769   15751 miner.go:109] Can not start mining operation due to network sync (starts when finished)
    I0829 19:33:28.809242   15751 miner.go:109] Can not start mining operation due to network sync (starts when finished)
    I0829 19:33:28.828618   15751 chain_manager.go:661] imported 256 block(s) (0 queued 0 ignored) including 144 txs in 3.655024716s. #159006 [7118fd57 / 1b191ac1]
    I0829 19:33:30.413771   15751 miner.go:119] Starting mining operation (CPU=0 TOT=1)
    I0829 19:33:31.209846   15751 worker.go:540] commit new work on block 159037 with 19 txs & 0 uncles. Took 795.246035ms
    I0829 19:33:39.042680   15751 chain_manager.go:661] imported 256 block(s) (0 queued 0 ignored) including 448 txs in 10.212744298s. #159262 [4a601ac3 / 22298749]

**Pull request merged**